### PR TITLE
feat: Implement cache first strategy for Perps Activity View

### PR DESCRIPTION
**Pull request by gambinish**
## **Description**

Improves Perps Activity tab performance by implementing a cache-first strategy to loading the page to prevent unnecessary API calls to Hyperliquid and long time to render on the client.

## **Changelog**

CHANGELOG entry: Add cache first strategy to perps activity view

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/TAT-2212

## **Screenshots/Recordings**

https://github.com/user-attachments/assets/eaa1aea3-2243-4bcc-a541-6a87f90a2856

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> Introduces a cache-first, progressive loading flow for Perps transaction history to reduce API calls and improve perceived performance.
> 
> - Adds provider-side transaction history cache (fills, orders, funding) in `HyperLiquidProvider` with `skipCache` support, cache key helpers, and cache clearing on account/disconnect; updates `getOrderFills`, `getOrders`, and `getFunding` to use/return cached data
> - Refactors `usePerpsTransactionHistory` to a two-phase fetch (Phase 1: fills+funding, Phase 2: orders in background), exposes `ordersLoaded`, handles race conditions via fetch versioning, rebuilds when user history arrives, merges with live fills, and passes `skipCache` on refresh
> - Updates `PerpsTransactionsView` to respect `ordersLoaded` (show skeleton only for Orders until loaded), adjust performance measurement conditions, and keep a single source of truth for filtered views
> - Mounts Perps providers in `ActivityView` only when the tab is active to prevent unnecessary polling
> - Adds `transactionsHistoryConfig` constants (FlashList tuning, funding lookback) and extends controller types with `skipCache`
> - Expands tests across hooks, provider, and view to cover caching, progressive loading, enrichment, deduping, errors, and edge cases
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 0e4a90202104308e487e004a5730a38c39474ab8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by abretonc7s — CHANGES_REQUESTED** [review]
The PR description mentions "slow load times" but this        
  implementation still fetches all transactions upfront (lines  
  97-108 in usePerpsTransactionHistory.ts), then slices         
  client-side. The "infinite scroll" just progressively reveals 
  already-fetched data from memory.                             
                                                                
  Why this may not help:                                        
                                                                
  1. FlashList already virtualizes rendering - it only renders  
  visible items regardless of array size. So limiting displayed 
  items doesn't improve render performance.                     
  2. The actual bottleneck is the initial fetch + transforms -  
  parallel API calls for all fills, orders, funding, plus heavy 
  BigNumber calculations in aggregateFillsByTimestamp().        
  3. Loaded ≠ Rendered - having 100 items in state is fine;     
  FlashList handles it. The issue is fetching/processing all    
  data before showing anything.                                 
                                                                
  API capabilities (verified from SDK                           
  @nktkas/hyperliquid@0.27.1):                                  
                                                                
  | API              | Pagination                         |     
  |------------------|------------------------------------|     
  | historicalOrders | ❌ None (only user param)          |     
  | userFillsByTime  | ✅ Time-based (startTime, endTime) |     
  | userFunding      | ✅ Time-based (startTime required) |     
                                                                
  Suggestions:                                                  
                                                                
  1. For fills (largest dataset): Use userFillsByTime with      
  actual time-based pagination - fetch recent fills first, load 
  older on scroll.                                              
  2. For orders: API doesn't support pagination, but consider if
   we need ALL historical orders or just recent ones.           
  3. Alternative: If WebSocket provides initial snapshot data,  
  use that for instant first render, then backfill with REST.   
  4. Simplest option: Remove the pagination complexity entirely 
  - just fetch data and let FlashList handle rendering. The     
  PAGE_SIZE slicing adds complexity without addressing the fetch
   bottleneck.

File: app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts [hunk 1]
```diff
@@ -130,34 +129,137 @@ export const usePerpsTransactionHistory = ({
...
+
+            // Store orders in ref
+            ordersRef.current = orders;
+
+            // Re-build transactions with orders (now enriched with TP/SL pills)
+            const phase2Transactions = buildTransactions(
+              fillsRef.current,
+              orders,
+              fundingRef.current,
+              true,
```
[the remark this answers is not in the document]
Comment (gambinish) [reply app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts:270]: https://github.com/MetaMask/metamask-mobile/pull/24089/commits/679bed820bd75ad95962287cef93fbccf05a90d3

File: app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts [hunk 2]
```diff
@@ -130,34 +129,137 @@ export const usePerpsTransactionHistory = ({
...
+              transactionCount: phase2Transactions.length,
+            });
+          })
+          .catch((err) => {
+            // Orders failed - fills/funding still work, just no TP/SL pills or Orders tab
+            DevLogger.log('Failed to fetch orders for enrichment:', err);
+            // Don't set error - Phase 1 data is still valid
+            // Mark orders as "loaded" so Orders tab shows empty state instead of spinner
+            setOrdersLoaded(true);
+          });
```
[the remark this answers is not in the document]
Comment (gambinish) [reply app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts:289]: https://github.com/MetaMask/metamask-mobile/pull/24089/commits/679bed820bd75ad95962287cef93fbccf05a90d3

File: app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts [hunk 3]
```diff
@@ -130,34 +129,137 @@ export const usePerpsTransactionHistory = ({
...
+    fundingRef.current = [];
+    ordersRef.current = [];
+
     // Fetch user history first, then fetch all transactions
     const freshUserHistory = await refetchUserHistory();
     userHistoryRef.current = freshUserHistory;
-    await fetchAllTransactions();
-  }, [fetchAllTransactions, refetchUserHistory]);
+    await fetchTransactions(true); // skipCache = true for refresh
+  }, [fetchTransactions, refetchUserHistory]);
```
[the remark this answers is not in the document]
Comment (gambinish) [reply app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts:322]: https://github.com/MetaMask/metamask-mobile/pull/24089/commits/679bed820bd75ad95962287cef93fbccf05a90d3

File: app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts [hunk 4]
```diff
@@ -130,34 +133,173 @@ export const usePerpsTransactionHistory = ({
...
 
+  // Initial fetch on mount
   useEffect(() => {
     if (!skipInitialFetch && !initialFetchDone.current) {
       initialFetchDone.current = true;
-      refetch();
+      fetchTransactions(false); // Use cache on initial load
     }
-  }, [skipInitialFetch, refetch]);
+  }, [skipInitialFetch, fetchTransactions]);
```
[the remark this answers is not in the document]
Comment (gambinish) [reply app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts:330]: https://github.com/MetaMask/metamask-mobile/pull/24089/commits/2408f677adb3e914db36dcd4b0bb31a9037f725d

File: app/components/UI/Perps/Views/PerpsTransactionsView/PerpsTransactionsView.tsx [hunk 5]
```diff
@@ -354,18 +356,24 @@ const PerpsTransactionsView: React.FC<PerpsTransactionsViewProps> = () => {
...
-    [isConnecting, transactionsLoading],
-  );
+  // For Orders tab, also show skeleton while orders are loading in background
+  const isInitialLoading = useMemo(() => {
+    // Show loading if we're connecting or if transaction data is loading
+    if (isConnecting || transactionsLoading) {
+      return true;
+    }
+    // For Orders tab, show loading until orders are loaded
+    if (activeFilter === 'Orders' && !ordersLoaded) {
```
Comment (michalconsensys) [inline app/components/UI/Perps/Views/PerpsTransactionsView/PerpsTransactionsView.tsx:366]: Why is it only for Orders tab?

File: app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts [hunk 6]
```diff
@@ -130,34 +134,200 @@ export const usePerpsTransactionHistory = ({
...
+        }
+        const errorMessage =
+          err instanceof Error
+            ? err.message
+            : 'Failed to fetch transaction history';
+        DevLogger.log('Error fetching transaction history:', errorMessage);
+        setError(errorMessage);
+        setTransactions([]);
+        setIsLoading(false);
+      }
```
[the remark this answers is not in the document]
Comment (michalconsensys) [reply app/components/UI/Perps/hooks/usePerpsTransactionHistory.ts:303]: @gambinish does it need to be addressed?

**Comment by gambinish** [comment]
I don't think that this is required anymore. There's a commit that fixes the performance issue, and I've made a separate PR with only that fix here: https://github.com/MetaMask/metamask-mobile/pull/24271
